{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we will provide code on how to recreate the plots for the cystic fibrosis results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from biom import load_table\n",
    "from skbio.stats.composition import clr, clr_inv, centralize, closure\n",
    "from skbio.stats.ordination import OrdinationResults\n",
    "from scipy.spatial.distance import euclidean\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '../data/CF'\n",
    "\n",
    "microbes = load_table(os.path.join(data_dir, 'otus_nt.biom')).to_dataframe().T\n",
    "metabolites = load_table(os.path.join(data_dir, 'lcms_nt.biom')).to_dataframe().T\n",
    "\n",
    "# align the samples\n",
    "ids = set(microbes.index) & set(metabolites.index)\n",
    "microbes = microbes.loc[ids]\n",
    "metabolites = metabolites.loc[ids]\n",
    "\n",
    "# to keep consistent with the mmvec analysis, microbes with less than 10 reads total are dropped\n",
    "microbes = microbes.loc[:, (microbes>0).sum(axis=0) > 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "arrow-metadata.txt\r\n",
      "emperor.qzv\r\n",
      "\u001b[34mlatent_dim_3_input_prior_1.00_output_prior_1.00_beta1_0.85_beta2_0.90\u001b[m\u001b[m\r\n",
      "metabolites-pca.results\r\n",
      "microbes-pca.results\r\n",
      "model.txt\r\n",
      "model_run13_CF_dim3_lr1e-4_beta10.9_beta20.99_sub100_batch10.txt\r\n",
      "omics-biplot.results\r\n",
      "ordination.qza\r\n",
      "ordination.txt\r\n",
      "ordination_run13_CF_dim3_lr1e-4_beta10.9_beta20.99_sub100_batch10.txt\r\n",
      "point-metadata.txt\r\n",
      "ranks.csv\r\n"
     ]
    }
   ],
   "source": [
    "!ls ../results/cf_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = pd.read_csv(\n",
    "    '../results/cf_output/model_run13_CF_dim3_lr1e-4_beta10.9_beta20.99_sub100_batch10.txt',\n",
    "    index_col=0\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Shape of passed values is (462, 3), indices imply (138, 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   1652\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1653\u001b[0;31m         \u001b[0mmgr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBlockManager\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1654\u001b[0m         \u001b[0mmgr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, blocks, axes, do_integrity_check)\u001b[0m\n\u001b[1;32m    113\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mdo_integrity_check\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 114\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_verify_integrity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36m_verify_integrity\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    310\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mblock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_verify_integrity\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mblock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mmgr_shape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 311\u001b[0;31m                 \u001b[0mconstruction_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtot_items\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    312\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mtot_items\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mconstruction_error\u001b[0;34m(tot_items, block_shape, axes, e)\u001b[0m\n\u001b[1;32m   1690\u001b[0m     raise ValueError(\"Shape of passed values is {0}, indices imply {1}\".format(\n\u001b[0;32m-> 1691\u001b[0;31m         passed, implied))\n\u001b[0m\u001b[1;32m   1692\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (462, 3), indices imply (138, 3)",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-3d1d424e13f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m \u001b[0mpc_ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'PC0'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PC1'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'PC2'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m \u001b[0mU\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mu\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmicrobes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpc_ids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m \u001b[0mV\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmetabolites\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpc_ids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m \u001b[0meigvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpc_ids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    422\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m                 mgr = init_ndarray(data, index, columns, dtype=dtype,\n\u001b[0;32m--> 424\u001b[0;31m                                    copy=copy)\n\u001b[0m\u001b[1;32m    425\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    426\u001b[0m         \u001b[0;31m# For data is list-like, or Iterable (will consume into list)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36minit_ndarray\u001b[0;34m(values, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    165\u001b[0m         \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmaybe_infer_to_datetimelike\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    166\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 167\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mcreate_block_manager_from_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    168\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   1658\u001b[0m         \u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'values'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1659\u001b[0m         \u001b[0mtot_items\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1660\u001b[0;31m         \u001b[0mconstruction_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtot_items\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1661\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1662\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/mmvec/lib/python3.7/site-packages/pandas/core/internals/managers.py\u001b[0m in \u001b[0;36mconstruction_error\u001b[0;34m(tot_items, block_shape, axes, e)\u001b[0m\n\u001b[1;32m   1689\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Empty data passed with indices specified.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1690\u001b[0m     raise ValueError(\"Shape of passed values is {0}, indices imply {1}\".format(\n\u001b[0;32m-> 1691\u001b[0;31m         passed, implied))\n\u001b[0m\u001b[1;32m   1692\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1693\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (462, 3), indices imply (138, 3)"
     ]
    }
   ],
   "source": [
    "from scipy.sparse.linalg import svds\n",
    "\n",
    "microbes = model.loc[model.embed_type=='microbe']\n",
    "metabolites = model.loc[model.embed_type=='metabolite']\n",
    "pc_ids = ['PC0', 'PC1', 'PC2']\n",
    "microbes = microbes[['feature_id', 'axis', 'mean']].pivot(index='feature_id', columns='axis')\n",
    "metabolites = metabolites[['feature_id', 'axis', 'mean']].pivot(index='feature_id', columns='axis')\n",
    "microbes.columns = microbes.columns.droplevel()\n",
    "metabolites.columns = metabolites.columns.droplevel()\n",
    "\n",
    "microbes['ones'] = 1\n",
    "metabolites['ones'] = 1\n",
    "microbes = microbes[['bias', 'ones'] + pc_ids]\n",
    "metabolites = metabolites[['ones', 'bias'] + pc_ids]\n",
    "\n",
    "logits = microbes @ metabolites.T\n",
    "logits = logits - logits.mean(axis=0)\n",
    "logits = logits - logits.mean(axis=1).values.reshape(-1, 1)\n",
    "\n",
    "u, s, v = svds(logits.values.T, k=3)\n",
    "scale = 1.   # aspect ratios between metabolite points and microbe arrows\n",
    "#n = v.shape[1]\n",
    "#u = u * np.sqrt(n-1) / scale \n",
    "#v = scale * np.diag(s) @ v / np.sqrt(n-1)\n",
    "\n",
    "pc_ids = ['PC0', 'PC1', 'PC2']\n",
    "U = pd.DataFrame(u, index=microbes.index, columns=pc_ids)\n",
    "V = pd.DataFrame(v.T, index=metabolites.index, columns=pc_ids)\n",
    "eigvals = pd.Series(s, index=pc_ids)\n",
    "proportion_explained = pd.Series(s**2 / np.sum(s**2), index=pc_ids)\n",
    "res = OrdinationResults(\n",
    "    short_method_name='Multiomics biplot',\n",
    "    long_method_name='Multiomics biplot',\n",
    "    samples=U, features=V, eigvals=eigvals, \n",
    "    proportion_explained=proportion_explained\n",
    ")\n",
    "res.write('../results/cf_output/ordination.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAscAAAFgCAYAAABXB9TlAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3deZxddXn48c8DYYmgEyCAEJZoUURFRVPcCoVWRMFURK2tC8bWBsHtp/3h2lZcscVisVVsbDUuKAIFSgR/oQJxwSUmAiJLKiBbQGQdWQKB8Pz+ON9LLjf3zsyd3G3mft6v13nduef7Pec+98zMM8987znfE5mJJEmSJNik3wFIkiRJg8LiWJIkSSosjiVJkqTC4liSJEkqLI4lSZKkwuJYkiRJKiyOtVEi4rqIyLIcN07fk+v6Lmtoy4gY2HkFI2JuifG6SWw70O9N0mDpVF4dJBFxbInx2B69Xu0Yzu3F601GRCwrMR7Q71j0WBbH6qQjImLTZg0R8QTgVT2OR5Kmup7m1V4XsdIgsjhWp6wAdgYOatH+F8BM4Oct2vcqy3Q0nd+bpO7Z2LyqwXYE1d+G5f0ORI9lcaxOWVweF7RoXwCsA77erDEzr8rMqzoe1QCYzu9NUlctLo8LWrQvYIy8qsGWmTeUvw/39zsWPZbFsTrlZ8AVwCsjYlZ9Q0TsCbwQWArc0mzjsc7LjYjNImJhRFwYEXdGxIMRcUNEfCci3tDQ99FzuCJi/4g4JyJuj4hHIuKwhn2+IyJ+FhG/j4g1EXFlRHw6IrYd641GxIyI+EDp/0BE3BoRX42I3Sbx3mZHxD9GxFUlht9HxE8j4uiImNGk/4Kyv8URsU1EfK4ci1r8b6vr+4yIOLXEtyYilkfEwWO8r60i4n0R8fO6Y3J5+Zh167GOiaSu2Ki8WvptFxGfiIjLIuLeiLgvIn4REe+JiM0a+ibwkfL0I3XnMj/mNIuIeH5EHB8RK0p+WRsRN0fE6RHxgvHeVEQ8KSK+UbZ9oOSZv22W80r/iIg3lfx+V9nmmoj4fETsOt7rNdnfVhHx4Yi4tO6YXBIRH4qIxzXpf0A5BssiYsuI+HhEXF1y5LUR8XdRTn2JiF0j4j8jYnWJ87KIeGOLOMY85zgiDo6IM8qxXRsRv42IiyLi/RExs67fphHxtoj4cUSMlr63lu/zP0fE9u0eo6GXmS4uk16A64AE5gHHlK+PauhzXFn/mrIksKyhT1Y/jhvsfxvgx6X9AeB84FvA94G7gOsa+i8rfT9PNaJyWen/P8Chpc+WwIWl333Ad4BTqf7AZHlPT27Y79y6tv8CHqT6o3QKcGNpuxXYs8l7aPXe9gBuKO23lBi+U2LKEvMWDdssKG1nAVcBN5ftzgceKm3vp/qjeQ/wqxLjytL2MLB/k1h2AS4vfX5X3tt/l/0ncCmwTb9/3lxchmHpYF7dG1hd2m4s+eVc4I6y7nvA5nX9FwOXlLZLyvPaclhdv++VXHIpsAQ4veTaWo55bZP3dGxp/2p5/dXAt4HvUuX2BM4ENmnYLoCTS/ta4LyS064t6+4A/nCMYzi3Yf1s4Jel7c7ymmdQ/T2pve9tG7Y5oLT9GPhh2e6MEnstX58E/AHw2xLbKaVvluUNTWJcVtoOaPKeT6rb9udUf8eWsv5vxtyG71sC95fj883yeE1Z/4J+/0xPtaXvAbhM7YXHJvEnlsT4s7r2TUsSvAPYnPaL4/+uS0o7N7RtCby8YV0t2SSwsEXM/1TarwTm1K2fSVX4JvCThm3m1u33VuDpdW2bU32smcDyJq/X6r0tL22nAlvWrd8VWFXajmvYZkFdHKc1bPfysv6e8n3524Ztjy/t5zesD9b/A/KvwOMajkntvS3u98+bi8swLJ3Iq+V3t1ZAfhCYUde2LdU/3wkc2/DaxzZb39DnZcCOTdbPpypg76jPIw37Tapiuj53PQW4qbQd3bDd0WX9b4FnNByDz7F+0KJxIKF2DOc2rD+1rP8BMKtu/TbARaXtWw3bHFAX+w+Bkbq2Z5f3vI5qlP9fgE3r2t9etru6yfFaRvPi+D117/kFDW0BHFiLAdi99L2hxffkOcAO/f6ZnmpL3wNwmdpLfRIvz79Tnu9VntcKtn8tzydcHJdf6lqxt/0E46klm/NatM8s+0vgoCbts4F7S/uL69bPrUuO72iy3Qhwd+N2Y7y3/cr639MwStFw3H7PY/+ILKhbP7vJdrVRnx83adu2tD0IbNbktX5Cw6hNad+K6h+Ch3D02MWl60sn8ipwVFn37RavsTNVUXcbEHXrj2Wc4nic2GujvIc2rK/t936aFGvAW0r7rxvW10Y//6bJNpuzfiT1DQ1ttWM4t27d7sAjVIXs05rs7+mlbR2wa936A8q+1tW+Bw3bncX6In3zhrYZrB+p362hbRkNxXHpf1tZ/7IJHO8/LH3P6vfP7XRaPOdYnba4PC5oeFxM+15WHv87M29rc9szWqx/HrA1cHNm/k9jY2beTvUxIVQJsZlvNNlulOoP2Fjb1fvj8rgkM+9ssr/vUp1q8fgSc6MVJdZGV5fH/9dkn3eyfqRpu7qmQ8rjf2XmI022u4/qqvkZVIlYUm8tLo8LGh4X01rt9/q0Zo2ZeTPwa6oBgae0G1C5XmJBRHwmIv6jXAexGHhm6fLUFpuel5m/a7L+ZKrCdY+ImFNeYxfgyWX9BhcdZubash1MLO/uRzXy+tNscpF0Zl5B9YneJsD+Tba/PjOvbLK+lncvKDHV7/Nh4Dfl6c4TiHEe1ffkpszcII83cRXVgM+h5Zzp3SewjcZhcaxOO5uqAHtTRMwGXglclpkrJ7Gv2i/5ZGZ6uL7F+jnl8Tct2qEaqajvW+/uzLy7xXbXlcddxg5twnFcO0YcN7XY5t4Jtm9Zt+7J5fH4hgtwHl1Y/4fWCzuk3ptMXq39Xp82xu/100uftn6vI+JIqhz7FeBvgb8G3lyWZ5VuT2ixedOcV4rK2oWFtRxay323ZOYDLfY3Vr5utLH5v5N5t5Xa371VE+hLZt4D/BWwBvgkcF1E3BQRp5V/XibymmrQ9MpQabIyc21EfBN4J1Xi3KI89tqaFuujPOYY28YYbRMx1r47FccGI7xttter3WDg+6wv8Ftp9U+HpC6ZZF6t/V6fAzT7lKneHRONJSLmUV0s9jDVxYJLqIrC+zMzI+JTVOc4b0wereXFTufrQcq7HZOZp0fE96j+adofeDHrL9Q8NiL2y8wb+xHbVGVxrG5YTJXEX0GVQE8es3drtUJszw7EVFP7z/5JY/Spta1u0jYrIkbKaRSN5pbHm9uI48lj9Bkrjk6qJc3TMvPzXX4tSZOzmPby6o1UufOkzDyng3G8hqqA/FxmfqZJ+x7jbD+32cqI2BzYqTyt5dBantw5IrbIzAebbNpOnhykvNvKpP7ulU80v1oWIuIPgC9RXbz3j8DrOxjjtOdpFeq4zPwF8COq0YjTWpxfNhFLy+Mry0eJnbCS6iOuORHxp42NEbEd1RXXUF0s0cwbGldExAjVH62xtqv3/fI4PyK2abK/g6n+UNxbYu6m75bH13b5dSRN0iTy6mR/r2vnzLYaPKvNA7/BSGSZT7fV3fxqXtpi3t2/pKpJrsnMmwDK47Vl/QZzBZd5mmtF37JxXhfWT632gojY4JzoiNgLeD7VCPAPJrC/blhJNdK/S4wxL/14MvMaqtMsoJpRQ22wOFZXZOZ+mTk7Myf932pmXkz1kd3jgTMjYqf69jIZ+8vb3Oca4Ivl6Yn1+yznZp1EdcHeTzPzoha7+YeSRGvbbQacSDVjxcrM/NEE4vgh1dyVjwc+HxFb1O1vDtV0QAD/Nsa5dp1yFlVC/uOI+GI0uQlKRDw5It7e5TgkjaHNvLqIqoB9c0R8tMXNLZ4ZEW9pWF0bMW11y/vaNSBHRN3NgSLi8cCXgVlNt1rvccC/NeS8PwA+Xp6e2ND/hPL48Yh4Wt02m1JNy7k71Wjr6eO8Lpl5PdV0nZsA/14GNWr7mwX8e2k7tV+nIWTmQ1RzWAN8JSL2rW+PygG12CNin4h4Xf1NQerUBno8Ha5NnlahQbeAauaFPwKujYgfUU1zszPVf8OjtPiYbgx/T3VF8AHAryPiAqpzlPejGq29gSajw8UNVIXkJWW7UaobbuxG9d/+EW3E8Xqqm5H8JVVh+iOqPxwHUk2fdj7V9EddlZm1uweeCxwJvD4iLqX6CHI21Xt7KtV0bp52IU0BmXlvRBxKNYvOPwDviIhfUs2duyPV6QNzqe7CV3/+8lKq6dYOj4gfUF2gtg44OzPPLn3/D/Bc1ufkoDrXdS1VgfxXY4T2deBQ4JqIuIhqMOJPqC5WW8KGOeYLVOfQ/iVwaURcSHXDjn2pTo+4i+qmI81OuWjmKOBpVPn/2ohYVtYfSDXX8aVUcxP302ep/jl5K/DTiFhBNSPGtlQXUe5K9f0bpfrn4BTg/oj4BdU/RJsD+1Adn3uovv9qgyPHGmhl+rH9qM61+wVVQjycKjH8EPjAJPb5APBS4F1Uk7YfSHUhw++pRiKem5nXttoc+HOqUY4nA4dRzZ38Daq7NF3RRhxXUyWw46lOn3glVcK+HHgH1Q1OJprwN0r5+HLf8roXA88AXk01LdM9wGeojrukKSIzL6OaPeJDVNO2PZfq9/ipVEXyx4GFDdv8luoUsWVl2zdTzUbx3NJ+F9XgwiKqvHVoeX5G6TPeiOu1VFNC/ogq976E6kLg9wGvbpxOMjOTarDiCKpC/gXlPWxC9UnfszPz5xM8JLXpOl9INUiymmrO6JeXuD9MNU/9BtNr9lJW/oZq5Pdcqn9iXkM19//1VMfqt6X7T6kugPwB1Swfh1Ed0/uBfwb2zswVvYx/Oojq506SJEmSI8eSJElSYXEsSZIkFRbHkiRJUmFxLEmSJBVDO5Xb7Nmzc+7cuf0OQ9IQWrly5e2Z2exGCEPFPCypn1rl4qEtjufOncuKFc5uIqn3IsJJ+TEPS+qvVrnY0yokSZKkwuJYkiRJKiyOJUmSpMLiWJIkSSosjiVJkqRiyhfHEXF2RNwVEaf3OxZJGlbmYknTxZQvjoHPAkf0OwhJGnLmYknTwpQvjjPzQuCefschScPMXCxpuuhrcRwR+5eP4lZHREbEgiZ9jo6I30TEAxGxMiL260OokjRtmYslab1+jxxvDfwKeDewprExIl4HnAh8CtgH+DHw3YjYrZdBStI0Zy6WpKKvxXFmnpuZH8rM04FHmnR5L7A4M7+UmVdm5juBW4CjehqoJE1j5mJJWq/fI8ctRcTmwPOA8xqazgNeNMl9LoyIFRGx4rbbbtvYECVp2ut0LjYPSxp0A1scA7OBTYFbG9bfCjyx9iQivgecBhwSETdFxAtb7TAzF2XmvMyct/3223cjZkmabjqai83DkgbdjH4HMAHZ8Dzq12XmS3objiQNJXOxpKEwyCPHtwPrqBuZKHZgwxEMSVJ3mIslDZWBLY4zcy2wEjiooekgqiulJUldZi6WNGz6elpFRGwN7FGebgLsFhHPAe7MzBuAE4CvR8Ry4CLgbcDOwBf7Ea8kTUfmYklar9/nHM8DLqx7/tGyfBVYkJnfjojtgL8DdqKah/OQzLy+55FK0vRlLpakoq/FcWYuo7qoY6w+XwC+0JOAJGkImYslab2BPedYkiRJ6jWLY0mSJKmwOJYkSZKKoSuOI2J+RCwaHR3tdyiSNJTMw5IG2dAVx5m5JDMXjoyM9DsUSRpK5mFJg2zoimNJkiSpFYtjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKkYuuI4IuZHxKLR0dF+hyJJQ8k8LGmQDV1xnJlLMnPhyMhIv0ORpKFkHpY0yIauOJYkSZJasTiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZKKoSuOI2J+RCwaHR3tdyiSNJTMw5IG2dAVx5m5JDMXjoyM9DsUSRpK5mFJg2zoimNJkiSpFYtjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkoqhK44jYn5ELBodHe13KJI0lMzDkgbZ0BXHmbkkMxeOjIz0OxRJGkrmYUmDbOiKY0mSJKkVi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkoqhK44jYn5ELBodHe13KJI0lMzDkgbZ0BXHmbkkMxeOjIz0OxRJGkrmYUmDbOiKY0mSJKkVi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqZjR7gYR8SLgpcCOwL9k5qqI2Bp4FnB5Zo52OEZJUgNzsSR1x4RHjiNik4g4Gfgh8A/AQmBOaX4YOAc4quMRSpIeZS6WpO5q57SK9wF/Abwf2BuIWkNmPgCcCRza0egkSY3MxZLURe0UxwuAr2fmZ4DfNmm/EtijE0FJklpagLlYkrqmneJ4LnDRGO13AdtsVDSSpPHMxVwsSV3TTnF8L2Mn3D2A2zcuHEnSOMzFktRF7RTHFwFvaNYQEbOAtwAXdiIoSVJL5mJJ6qJ2iuNPAk+LiO8BLyvrnhkRfw2sAB4PfLrD8UmSHstcLEldNOF5jjNzeUS8FvgP4Gtl9WeprpS+A3h1Zl7e+RAlSTXmYknqrrZuApKZZ0fE7sDBwF5UyfjXwLmZeV8X4pMkNTAXS1L3tH2HvMxcA5xVFklSH5iLJak72rlD3rMj4sgx2o+MiGd1JixJUjPmYknqrnYuyDsWeNUY7X8GfGSjoumBiJgfEYtGR0f7HYokTcaxTPFcbB6WNMjaKY73BZaN0f594PkbFU0PZOaSzFw4MjLS71AkaTKmfC42D0saZO0Ux7OproRu5a7SR5LUPeZiSeqidorj3wFPH6P96cCdGxeOJGkc5mJJ6qJ2iuMLgLdGxNMaG8q6t5Y+kqTuMRdLUhe1M5Xbx6kuAlkZEV8CLgES2IcqGT9S+kiSusdcLEld1M4d8q6OiIOAxcC7qJJxlOargLdk5qqORyhJepS5WJK6q9075P0sIp4OPA94ClVCXgX8IjOzC/FJkhqYiyWpeyZzh7wEVpRFktQH5mJJ6o62i2OAiNgC2Jb1H+U9KjNv3tigJEnjMxdLUudNuDiOiE2A9wLvBHYZo+umGxuUJKk5c7EkdVc7I8efBN5PdcHHvzP2JPSSpO4wF0tSF7VTHB8BnAe83As+JKlvzMWS1EXt3ARkG+BMk7Ek9ZW5WJK6qJ3i+Apgp24FIkmaEHOxJHVRO8Xxx4AjI2JOt4KRJI3LXCxJXdTOOcfPBK4HroyI04HfAOsa+mRmHtep4CRJGzAXS1IXtVMcf6Lu6wUt+iRgQpak7jEXS1IXtVMcP6VrUUiSJspcLEldNOHiODOv6WYgkqTxmYslqbvauSBPkiRJmtbaOa2CiNgUmA88n2quzcbiOjPzyA7FJklqwlwsSd0z4eI4IrYBzgeeDQTVBR9RmrNunQlZkrrEXCxJ3dXOaRUfp5pC6G3AnlQJ+BBgb+A04OfADp0OUJL0GOZiSeqidorjVwBfy8wvAXeWdWsz8/LM/AvgQarJ6SVJ3WMulqQuaqc43glYXr5+uDxuWdd+JvDKTgTVb2ddvJoXf/oCnvSBc3jxpy/grItX9zskSaoZmlwsSf3QzgV5dwFbla/vAR4CdqlrfxDYtkNx9c1ZF6/mg2dcxpqHqhtOrb57DR884zIADtvHu7VK6ruhyMWS1C/tjBz/L7AXQGY+AlwKHBERm0XElsCbqG5jOqUdv3TVo4VxzZqH1nH80lV9ikiSHmMocrEk9Us7xfF5wGsiYovy/ATgRVTnvN0K7Auc2Nnweu/mu9e0tV6SemwocrEk9Us7p1UcB3wWWAuQmadExCPAG4F1wOmZeXLnQ+ytnWfNZHWTQnjnWTP7EI0kbWAocrEk9Us7t49eB9zXsO5U4NROB9VNETEfmL/HHns0bT/m4D0fc84xwMzNNuWYg/fsUYSS1Np0yMXj5WFJ6qehu310Zi7JzIUjIyNN2w/bZw7HHb43c2bNJIA5s2Zy3OF7ezGeJHXIeHlYkvqp5chxRLweIDO/Wf98PLX+U9lh+8yxGJY0EIY5F0tSP4x1WsU3gIyI0zNzbe05629T2kwCJmRJ6hxzsST10FjF8UEAJRk/+lyS1FPmYknqoZbFcWaeP9ZzSVL3mYslqbcmdEFeRGwdEasi4t3dDkiS1Jy5WJK6b0LFcWbeC+xIw/RBkqTeMRdLUve1M5XbcuB53QpEkjQh5mJJ6qJ2iuMPAK+LiDd1KxhJ0rjMxZLURe3cPvrTwB3A4oj4J+Aa4P6GPpmZB3cqOEnSBszFktRF7RTHT6eaO/Pm8nz3Jn1yoyOSJI3FXCxJXTTh4jgzd+lmIJKk8ZmLJam72jnnWJIkSZrWLI4lSZKkop1zjomIJwHvBp4PbMOGxXVm5p4dik2S1IS5WJK6Z8LFcUQ8A7gImAlcDTwVuAqYXZbfsP4CEUlSF5iLJam72jmt4mPAw8A+wB+Xde/IzB2AtwOPB/6ms+FJkhqYiyWpi9opjvcDFmXmFayfJigAMvMkYCnwj50NT5LUwFwsSV3UTnH8BKqP8ADWlset6tovAv6oE0FJkloyF0tSF7VTHN8K7AiQmfcA9wFPqWsfoc0L/CRJbTMXS1IXtZNALwXm1T3/IfCuiPgJVZH9duCXHYxNfXbWxas5fukqbr57DTvPmskxB+/JYfvM6XdY0rAzF0tSF7UzcnwK8MSImFme/z2wHVVi/n75+sOdDU/9ctbFq/ngGZex+u41JLD67jV88IzLOOvi1f0OTRp25mJJ6qIxR44jYovMfBAgM78JfLPWlpkry5RCrwbWAedk5tXN96Sp5vilq1jz0LrHrFvz0DqOX7rK0WOpx8zFktQ7451WcUtEfAv4cmaubGzMzOuBE7oSmfrq5rvXtLVeUleZiyWpR8Y7rWIUOApYHhGXRMQ7I2LbHsSlPtt51sy21kvqKnOxJPXImMVxZj4JeAnVR3h7ACcCqyPilIh4aURED2JUHxxz8J7M3GzTx6ybudmmHHOwd6SVes1cLEm9M+4FeZl5QWa+CdgJeBtwCfDnwHeB6yLioxHxpO6GqV47bJ85HHf43syZNZMA5syayXGH7+35xlKfmIslqTciM8fv1bhRxJ7AXwNvBJ4IPEJ1lfR/lotFBt68efNyxYoV/Q5D0hCKiJWZOW/8nuPuZ0rnYvOwpH5qlYvbmcrtUZm5KjPfB+wCzAf+BzgQ+NpGRSlJmjBzsSR13qSK4zr7An8GvLA8XztGX0lSd5iLJalD2r7FaEQ8EXgT8BZgTyCozn37T+DkjkYnSWrKXCxJ3TGh4jgiZlCNSrwFOLhsdzdwEtW5bRd3LUJJEmAulqReGO8Oec+iSsJvoLolKcCFVCMTZ9Tu2CRJ6h5zsST1zngjx5eUxxuBTwBfyczruhqRJKmRuViSemS84vh0qpGJ83Iyc75JkjrBXCxJPTJmcZyZf96rQCRJzZmLJal3NnYqN0mSJGnasDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZKKoSuOI2J+RCwaHR3tdyiSNJTMw5IG2dAVx5m5JDMXjoyM9DsUSRpK5mFJg2zoimNJkiSpFYtjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSiilfHEfEIRGxKiJ+HRFH9zseSRpG5mJJ08WMfgewMSJiBnAi8CfAHcCKiDgzM2/pb2SSNDzMxZKmk6k+crwvcEVm3piZ9wNnAq/oc0ySNGzMxZKmjb4WxxGxf0ScHRGrIyIjYkGTPkdHxG8i4oGIWBkR+9U17wzcWPf8JmBOl8OWpGnFXCxJ6/V75Hhr4FfAu4E1jY0R8Tqqj+o+BewD/Bj4bkTsVuvSZJ/ZnVAladoyF0tS0dfiODPPzcwPZebpwCNNurwXWJyZX8rMKzPzncAtwFGlfTWwa13/XYCbW71eRCyMiBURseK2227r0LuQpKmtl7nYPCxp0PV75LiliNgceB5wXkPTecCLytfLgWdExK4RMRN4FXBOq31m5qLMnJeZ87bffvtuhC1J00qnc7F5WNKgG9jiGJgNbArc2rD+VuCJAJn5MPAe4HzgMuCkzGw5cixJapu5WNJQmQpTuTWetxb16zJzCbCkpxFJ0vAxF0saCoM8cnw7sI4yMlFnBzYcwZAkdYe5WNJQGdjiODPXAiuBgxqaDqK6UlqS1GXmYknDpq+nVUTE1sAe5ekmwG4R8Rzgzsy8ATgB+HpELAcuAt5GNZ/mF/sRryRNR1w99wYAAAdfSURBVOZiSVqv3+cczwMurHv+0bJ8FViQmd+OiO2AvwN2opqH85DMvL7nkUrS9GUulqSir8VxZi6j+eTx9X2+AHyhJwFJ0hAyF0vSegN7zrEkSZLUaxbHkiRJUjF0xXFEzI+IRaOjo/0ORZKGknlY0iCLzMZ53YdDRNwG3A20k51nU835qfaN0N6xHhSDEHcvYujGa3Rinxuzj8lu2+52k8kLu2fm0N872Tzcc4OQzyZjEOLuVQydfp1O7a/XubgXeRha5eLMHNoFWNRm/xX9jnmqLu0e60FZBiHuXsTQjdfoxD43Zh+T3da80NvF4z24x3pQlkGIu1cxdPp1OrW/XufifueFoTutooG3Ou2dqXqsByHuXsTQjdfoxD43Zh+T3XYQvufDxOPdO1P1WA9C3L2KodOv06n99ToX9/V7PrSnVUxGRKzIzHn9jkPS4DAv9JbHW1KjTueFYR85bteifgcgaeCYF3rL4y2pUUfzgiPHkiRJUuHIsSRJklRYHEuSJEmFxbEkSZJUWBxLkiRJhcVxB0TEIRGxKiJ+HRFH9zseSf0XEWdHxF0RcXq/YxkW5mJJjSaTi52tYiNFxAzgSuBPgDuAFcCfZuYtfQ1MUl9FxIHA1sCbM/M1/Y5nujMXS2pmMrnYkeONty9wRWbemJn3A2cCr+hzTJL6LDMvBO7pdxxDxFwsaQOTycVDXxxHxP5lyH11RGRELGjS5+iI+E1EPBARKyNiv7rmnYEb657fBMzpctiSuqgDeUFtMhdLatSvXDz0xTHVUPuvgHcDaxobI+J1wInAp4B9gB8D342I3WpdmuzTc1WkqW1j84LaZy6W1KgvuXjoi+PMPDczP5SZpwOPNOnyXmBxZn4pM6/MzHcCtwBHlfbVwK51/XcBbu5q0JK6qgN5QW0yF0tq1K9cPPTF8VgiYnPgecB5DU3nAS8qXy8HnhERu0bETOBVwDm9i1JSL00wL6iDzMWSGnUzF1scj202sClwa8P6W4EnAmTmw8B7gPOBy4CTMtPRCmn6GjcvAETE94DTgEMi4qaIeGHvQpx2zMWSGnUtF8/oZJTTWON5a1G/LjOXAEt6GpGkfhsvL7ykt+EMBXOxpEYdz8WOHI/tdmAddf+BFDuw4X8qkoaDeaH3POaSGnUtL1gcjyEz1wIrgYMamg6iuiJS0pAxL/Sex1xSo27mhaE/rSIitgb2KE83AXaLiOcAd2bmDcAJwNcjYjlwEfA2qvk0v9iPeCV1n3mh9zzmkhr1Ky8M/e2jI+IA4MImTV/NzAWlz9HA+4CdqObbe09m/qBXMUrqLfNC73nMJTXqV14Y+uJYkiRJqvGcY0mSJKmwOJYkSZIKi2NJkiSpsDiWJEmSCotjSZIkqbA4liRJkgqLY0mSJKmwOJYkSZIKi2NNWxFxQERkWf6tRZ8dImJt6bOsxyFK0rRnLtZUY3GsYfAA8PqI2KJJ25uAAB7ubUiSNHTMxZoSLI41DM4EtgFe2aTtLcC5wIM9jUiSho+5WFOCxbGGwS+AS6mS76MiYl/gGcBXmm0UEfMi4syIuD0iHoyIVRHx4YiY0bifiFgcEf8bEfdHxD0RcVFEvKrJPheXjw1HIuKkiPhdRDxQ+j+/c29ZkgaOuVhTgsWxhsVXgJdGxC516/4K+B3wncbOEXEIcBHwVOCfgXcBPwE+BnyrofurgKcBpwLvBj4JbAucERGvbxHPUmCXsr/jgGcC50bE4yfz5iRpijAXa/BlpovLtFyAA4AE/i+wHdXHdR8qbTOBu4HPlOf3AsvK11sCvwV+AMxo2Od7yj4PqFu3VZPXfhywCriiYf3isv0XGta/tqw/st/HzcXFxaWTi7nYZaotjhxrKGTmHcDZwIKy6nBgBPhyk+4HATtSjXDMiojZtYXqnDiAl9bt+77a1xHxuIjYjiohXwDsFRFPaPIan214fkF5fEo770uSphJzsaaCGeN3kaaNrwDnRMQfUX2Mtzwzr2jSb6/y2CxZ1+xY+yIidgA+QXWRyQ5N+s4Cft+w7tr6J5l5R0RANaoiSdOZuVgDzeJYw2QpsBr4CHAgcFSLflEejwEuadHnZoCosuh5VEn8c8DPgVFgHdVFJ6+nybn9mblunNeWpOnKXKyBZnGsoZGZ6yLia8AHgTXAKS26/ro83peZ3xtnt88Cng18LDM/Ut8QEW/dmHglaToyF2vQWRxr2HwRWAtcm5mjLfospbpy+gMR8e3MvLO+MSJmUl0ccg/VqAQ0jDJExDOprpyWJG3IXKyBZXGsoZKZNwDHjtPnvog4AjgLWBURXwaupjpf7WlUF5C8ClgGXAlcDrwvImpXRT8VOBL4FfDcrrwRSZrCzMUaZBbHUhOZuTQi/hD4APBGYHvgLuAa4ATgl6Xfuog4FPgM8GZgK6pE/Gaqj/hMyJI0SeZi9UNkNa+fJEmSNPSc51iSJEkqLI4lSZKkwuJYkiRJKiyOJUmSpMLiWJIkSSosjiVJkqTC4liSJEkqLI4lSZKkwuJYkiRJKv4/VE/EmOfcm4AAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x360 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "titlesize = 22\n",
    "fontsize = 18\n",
    "labelsize = 14\n",
    "fig, ax = plt.subplots(1, 2, figsize=(10, 5))\n",
    "ax[0].scatter(microbes.mean(axis=0), microbes.var(axis=0))\n",
    "ax[0].set_xscale('log')\n",
    "ax[0].set_yscale('log')\n",
    "ax[0].set_xlabel('Mean', fontsize=fontsize)\n",
    "ax[0].set_ylabel('Variance', fontsize=fontsize)\n",
    "ax[0].set_title('Microbiome', fontsize=titlesize)\n",
    "ax[0].tick_params(axis='x', labelsize=labelsize)\n",
    "ax[0].tick_params(axis='y', labelsize=labelsize)\n",
    "\n",
    "ax[1].scatter(metabolites.mean(axis=0), metabolites.var(axis=0))\n",
    "ax[1].set_xscale('log')\n",
    "ax[1].set_yscale('log')\n",
    "ax[1].set_xlabel('Mean', fontsize=fontsize)\n",
    "ax[1].set_ylabel('Variance', fontsize=fontsize)\n",
    "ax[1].set_title('Metabolomics', fontsize=titlesize)\n",
    "ax[1].tick_params(axis='x', labelsize=labelsize)\n",
    "ax[1].tick_params(axis='y', labelsize=labelsize)\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's load the `ranks.csv` and the `omics-biplot.results` files that were created earlier. \n",
    "We'll also load the feature metadata mapping files, since we will use this information to paint the biplots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '../data/CF'\n",
    "ranks = pd.read_csv('../results/cf_output/ranks.csv', index_col=0)\n",
    "#biplot_res = OrdinationResults.read('../results/cf_output/omics-biplot.results')\n",
    "\n",
    "taxonomy = pd.read_table(os.path.join(data_dir, 'taxonomy.tsv'), index_col=0)\n",
    "ms1_md = pd.read_table(os.path.join(data_dir, 'lcms_annotations.txt'), index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biplot_res.samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll want to do some preprocessing of the feature metadata to ease manipulation.  \n",
    "Here, we'll just parse the taxonomies to make it easier to read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa = pd.DataFrame(taxonomy.Taxon.apply(lambda x: x.split(';')).values.tolist(), \n",
    "                    index=taxonomy.index, \n",
    "                    columns=['kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cleaning up the metabolomics annotations is a little more tricky - we need to do two things\n",
    "\n",
    "1. Match up the MS1 retention times and masses to the MS2 annotations\n",
    "2. Clean up the chemical names.\n",
    "\n",
    "I'm going to use a nearest neighbors approach regarding the retention times and masses.\n",
    "The masses are expected to be more accurate (within +/- 1) whereas the retention times are expected to slip more (+/- 10).\n",
    "\n",
    "We will need to extract the masses and retention ties from the ids.  Here the m/z is the first value and the retention time is the second value.\n",
    "\n",
    "# WARNING!!!\n",
    "\n",
    "As of now, there is no standardized protocol for labeling MS1 features.  Some MS1 bucket tables will be labeled with scan ids, or have the retention time and m/z flipped. So proceed with caution.\n",
    "\n",
    "That can be done as follows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial import KDTree\n",
    "\n",
    "ms_ids = ranks.columns\n",
    "mz = list(map(lambda x: float(x[1:].split('mz')[0]), ms_ids))\n",
    "rt = list(map(lambda x: float(x[1:].split('mz')[1]), ms_ids))\n",
    "df = pd.DataFrame({'measured_mz': mz, 'measured_rt': rt}, index=ms_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will create a database from the MS1 annotations, so that we will be matching the MS1 labels to the MS2 labels.\n",
    "We will only be looking at the closest neighbor and we will be recording their distances and original retention time and m/z."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lookup = KDTree(ms1_md[['Precursor_MZ', 'RT_Query']])\n",
    "d, i = lookup.query(df.values, k=1, eps=.1)\n",
    "\n",
    "df['nearest_distance'] = d\n",
    "df['scan'] = [ms1_md.index[k] for k in i]\n",
    "df = df.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will merge the MS1 and MS2 annotations together."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms1_md = ms1_md.reset_index()\n",
    "ms_md = pd.merge(df, ms1_md, left_on='scan', right_on='#Scan#')\n",
    "ms_md = ms_md.set_index('index')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have some sort of unified metabolite annotation metadata, we can now start digging into individual molecules.\n",
    "\n",
    "Since P. aeruginosa is the best understood microbe in this system, we'll start by focusing on its known molecules, \n",
    "and merging that information into our metabolite annotations metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '../data/CF'\n",
    "validation = pd.read_csv(\n",
    "    os.path.join(data_dir, 'validated-molecules.csv'), \n",
    "    index_col=0, header=None\n",
    ") \n",
    "validation.columns = ['rank', 'expert_annotation', 'shorthand']\n",
    "ms_md = pd.merge(ms_md, validation, left_index=True, right_index=True, how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will want to start cleaning up some of the chemical names.  This is mainly for the\n",
    "sake of visualization (since many of these names are just too darn long)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_md['shorthand'] = ms_md['shorthand'].fillna('None')\n",
    "\n",
    "def annotation_f(x):\n",
    "    if x['shorthand'] != 'None':\n",
    "        return x['shorthand']\n",
    "    if x['nearest_distance'] > 10 or len(x['Compound_Name']) > 30:\n",
    "        return 'Other'\n",
    "    if x['Compound_Name'] == 'Rha-Rha-C10-C10 (M+H)':\n",
    "        return 'Rhamnolipid'\n",
    "    if 'phosphocholine' in x['Compound_Name']:\n",
    "        return 'Phosphocholine'\n",
    "    if 'Palmitoylcarnitine' in x['Compound_Name']:\n",
    "        return 'Palmitoylcarnitine'\n",
    "    if 'phosphoethanolamine' in x['Compound_Name']:\n",
    "        return 'Phosphoethanolamine'\n",
    "    if 'butyrolactone' in x['Compound_Name']:\n",
    "        return 'Butyrolactone'\n",
    "    if 'indole' in x['Compound_Name'].lower():\n",
    "        return 'indole'\n",
    "    else:\n",
    "        return x['Compound_Name']\n",
    "    \n",
    "ms_md['shorthand'] = ms_md.apply(annotation_f, axis=1)\n",
    "ms_md.index.name = 'sampleid'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Before something else breaks, let's save all of the feature metadata files.\n",
    "We can use these files for visualization later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_md.to_csv('../results/cf_output/point-metadata.txt', sep='\\t')\n",
    "\n",
    "taxa.index.name = 'featureid'\n",
    "taxa.to_csv('../results/cf_output/arrow-metadata.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will make a custom biplot -- it is going to be hairy, so take a deep breath."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import euclidean\n",
    " \n",
    "axis0='PC0'\n",
    "axis1='PC1'\n",
    "# only plot the top 30 features\n",
    "features = biplot_res.features\n",
    "features['importance'] = features.apply(lambda x: euclidean(np.zeros_like(x), x), axis=1)\n",
    "features.sort_values('importance', inplace=True, ascending=False)\n",
    "features.drop(['importance'], inplace=True, axis=1)\n",
    "biplot_res.features = features[:30]\n",
    "\n",
    "# Now actually make the biplot\n",
    "fig = plt.figure(figsize=(8, 8))\n",
    "\n",
    "ridx = ms_md.loc[ms_md.shorthand == 'Rhamnolipid'].index\n",
    "idx = ms_md.loc[ms_md.shorthand != 'Rhamnolipid'].index\n",
    "\n",
    "rhamolipid_color = '#C918C1'\n",
    "other_color = '#B2B2B2'\n",
    "fermenter_color = '#00B899'\n",
    "anaerobe_color = '#C69600'\n",
    "\n",
    "arrow_width = .01\n",
    "other = plt.scatter(biplot_res.samples.loc[idx, axis0], \n",
    "                    biplot_res.samples.loc[idx, axis1], \n",
    "                    color=other_color, label='other')\n",
    "for i in biplot_res.features.index:\n",
    "    label = taxa.loc[i, 'genus']    \n",
    "    if label == ' g__Pseudomonas':\n",
    "        pseudomonas = plt.arrow(\n",
    "            0, 0, biplot_res.features.loc[i, axis0], biplot_res.features.loc[i, axis1], \n",
    "            color='r', width=arrow_width)        \n",
    "    elif biplot_res.features.loc[i, 'PC0'] > 0:        \n",
    "        pathogens = plt.arrow(\n",
    "            0, 0, biplot_res.features.loc[i, axis0], biplot_res.features.loc[i, axis1], \n",
    "            color=fermenter_color, width=arrow_width)\n",
    "    else:\n",
    "        anaerobes = plt.arrow(\n",
    "            0, 0, biplot_res.features.loc[i, axis0], biplot_res.features.loc[i, axis1], \n",
    "            color=anaerobe_color, width=arrow_width)\n",
    "\n",
    "rhamnolipids = plt.scatter(biplot_res.samples.loc[ridx, axis0], \n",
    "                           biplot_res.samples.loc[ridx, axis1], s=50,\n",
    "                           color=rhamolipid_color, zorder=5, label='Rhamnolipid')\n",
    "\n",
    "compounds = [\n",
    "    'Pyochelin', 'Phenazine-1-carboxylic Acid', \n",
    "    'HHQ', 'NHQ', 'C9:db-PQS, C9:DB-NQNO', \n",
    "    'PQS/HQNO', 'Pyocyanin'\n",
    "]\n",
    "\n",
    "interesting_taxa = [' g__Pseudomonas', ' g__Prevotella', ' g__Veillonella', \n",
    "                    ' g__Fusobacterium']\n",
    "\n",
    "\n",
    "compound_idx = [ms_md.loc[c, 'shorthand'] in compounds for c in ms_md.index]\n",
    "compound_idx = ms_md.loc[compound_idx].index\n",
    "subsamples = biplot_res.samples.loc[compound_idx]\n",
    "plt.scatter(subsamples[axis0], subsamples[axis1], marker='+',\n",
    "            color='k', s=100, zorder=10)\n",
    "\n",
    "for i in subsamples.index:\n",
    "    x, y = subsamples.loc[i, axis0], subsamples.loc[i, axis1]\n",
    "    label = ms_md.loc[i, 'shorthand']\n",
    "    dx = np.sign(x)*50\n",
    "    dy = np.sign(y)*50\n",
    "    plt.annotate(\n",
    "        label,\n",
    "        xy=(x, y), xytext=(dx, dy),\n",
    "        textcoords='offset points', ha='right', va='bottom',\n",
    "        bbox=dict(boxstyle='round,pad=0.5', fc='yellow', alpha=0.5),\n",
    "        arrowprops=dict(arrowstyle = '->', connectionstyle='arc3,rad=0'))\n",
    "\n",
    "for i in biplot_res.features.index:\n",
    "    x, y = biplot_res.features.loc[i, axis0], biplot_res.features.loc[i, axis1]\n",
    "    label = taxa.loc[i, 'genus']\n",
    "    if label not in interesting_taxa: continue\n",
    "        \n",
    "    dx = np.sign(x)*30\n",
    "    dy = np.sign(y)*30\n",
    "    plt.annotate(\n",
    "        label,\n",
    "        xy=(x, y), xytext=(dx, dy),\n",
    "        textcoords='offset points', ha='right', va='bottom',\n",
    "        bbox=dict(boxstyle='round,pad=0.5', fc='yellow', alpha=0.5),\n",
    "        arrowprops=dict(arrowstyle = '->', connectionstyle='arc3,rad=0'))\n",
    "\n",
    "plt.legend([pseudomonas, pathogens, anaerobes, rhamnolipids], \n",
    "           ['pseudomonas', 'pathogens', 'anaerobes', 'rhamnolipids'], \n",
    "           fontsize=12, loc=4)\n",
    "\n",
    "#plt.legend()\n",
    "plt.xlim([-4, 2])\n",
    "plt.ylim([-2.5, 2])\n",
    "plt.xticks([])\n",
    "plt.yticks([])\n",
    "plt.xlabel('PC1', fontsize=24)\n",
    "plt.ylabel('PC2', fontsize=24)\n",
    "plt.savefig('../results/custom-cf-omics.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, this looks pretty ugly right?\n",
    "\n",
    "This plot was ultimately tidied up in inkscape, since there are so many labels.\n",
    "This is another motivation to push these sorts of plotting capabilities to Emperor.\n",
    "\n",
    "Another thing worth doing is to compare and contrast the P. aeruginosa results between the conditional probabilities and the Pearson correlations.\n",
    "\n",
    "Let's retrieve the original microbe/metabolite data and match their samples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = '../data/CF/'\n",
    "\n",
    "microbes = load_table(os.path.join(data_dir, 'otus_nt.biom')).to_dataframe().T\n",
    "metabolites = load_table(os.path.join(data_dir, 'lcms_nt.biom')).to_dataframe().T\n",
    "\n",
    "# cast to dense dataframes and normalize to 1\n",
    "microbes = pd.DataFrame(microbes).apply(lambda x: x / x.sum(), axis=1)\n",
    "metabolites = pd.DataFrame(metabolites).apply(lambda x: x / x.sum(), axis=1)\n",
    "\n",
    "\n",
    "# align the samples\n",
    "ids = set(microbes.index) & set(metabolites.index)\n",
    "microbes = microbes.loc[ids]\n",
    "metabolites = metabolites.loc[ids]\n",
    "\n",
    "# to keep consistent with the minstrel analysis, microbes with less than 10 reads total are dropped\n",
    "microbes = microbes.loc[:, (microbes>0).sum(axis=0) > 10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's calculate the Pearson correlation coefficients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import spearmanr, pearsonr\n",
    "\n",
    "# extract the P. aeroginosa deblurred sequence\n",
    "pseudomonas = taxa.loc[(taxa['genus'] == ' g__Pseudomonas').values].index[0]\n",
    "\n",
    "# extract the P. aeroginosa ranks / probabilities.\n",
    "pranks = ranks.loc[pseudomonas].sort_values()\n",
    "\n",
    "pear_ranks = pd.Series(index=ms_md.index)\n",
    "pear_pvals = pd.Series(index=ms_md.index)\n",
    "pranks = pranks.loc[pear_ranks.index]\n",
    "\n",
    "for i in ms_ids:\n",
    "    r = pearsonr(list(microbes[pseudomonas].values), list(metabolites[i].values))\n",
    "    pear_ranks.loc[i] = r[0]\n",
    "    pear_pvals.loc[i] = r[1]    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The comparison between the Pearson and Conditional probabilities is given below.\n",
    "The conditional probabilities here are log transformed and centered around the mean log.\n",
    "\n",
    "The dots are all sized by the -log pvalue from Pearson correlation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "ax.scatter(pranks, pear_ranks, s=-2*np.log(pear_pvals))\n",
    "ax.scatter(pranks.loc[ridx], pear_ranks.loc[ridx], color='r', \n",
    "            s=-2*np.log(pear_pvals.loc[ridx]), label='rhamnolipids')\n",
    "ax.set_xlabel('Conditional logits', fontsize=18)\n",
    "ax.set_ylabel('Pearson coefficients', fontsize=18)\n",
    "ax.set_title('P. aeruginosa molecular ranks', fontsize=18)\n",
    "ax.legend(loc=4, fontsize=14)\n",
    "\n",
    "compounds = pd.DataFrame({\n",
    "    'X211.0918mz161.2083': ['Pyocyanin', 30, -60],\n",
    "    'X325.0673mz300.6387': ['Pyochelin', -80, 40],\n",
    "    'X232.1329mz284.2357': ['PCA', 50, -25],\n",
    "    'X244.1681mz317.3049': ['HHQ', 30, -30],\n",
    "    'X258.1804mz348.1582': ['PQS/HQNO', -80, 10],\n",
    "    'X256.1688mz350.8455': ['PQS/HQNO', -20, 30],\n",
    "    'X273.2005mz375.3707': ['NHQ', -70, 20],\n",
    "    'X271.1934mz377.8070': ['NHQ', 80, -35],\n",
    "    'X272.2311mz381.5715': ['NHQ', -30, 40],\n",
    "    'X289.2395mz367.6799': ['NQNO', 10, -50],\n",
    "    'X295.1511mz367.9210': ['NQNO', 30, 30],\n",
    "    'X284.1965mz373.1668': ['NQNO', -30, 10]\n",
    "}, ['compound', 'dx', 'dy']).T\n",
    "\n",
    "subsamples = pd.merge(biplot_res.samples, compounds, left_index=True, right_index=True)\n",
    "\n",
    "for i in subsamples.index:\n",
    "    x, y = pranks.loc[i], pear_ranks.loc[i]\n",
    "    dx, dy = subsamples.loc[i, 'dx'], subsamples.loc[i, 'dy']\n",
    "    \n",
    "    label = subsamples.loc[i, 'compound']\n",
    "    #dx = np.sign(x)*50\n",
    "    #dy = np.sign(y)*50\n",
    "    ax.annotate(\n",
    "        label,\n",
    "        xy=(x, y), xytext=(dx, dy),\n",
    "        textcoords='offset points', ha='right', va='bottom',\n",
    "        bbox=dict(boxstyle='round,pad=0.5', fc='yellow', alpha=0.5),\n",
    "        arrowprops=dict(arrowstyle = '->', connectionstyle='arc3,rad=0'))\n",
    "    \n",
    "ax.xaxis.set_tick_params(labelsize=16)\n",
    "ax.yaxis.set_tick_params(labelsize=16)\n",
    "ax.xaxis.set_major_locator(MaxNLocator(4))\n",
    "ax.yaxis.set_major_locator(MaxNLocator(4))\n",
    "plt.tight_layout()\n",
    "plt.savefig('../results/rhamnolipids_comparison.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can zoom in on the Rhamnolipids to see how the ranks compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import rankdata\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "\n",
    "conditional_ranks = pd.Series(rankdata(-pranks), pranks.index)\n",
    "pearson_ranks = pd.Series(rankdata(-pear_ranks), pear_ranks.index)\n",
    "k = conditional_ranks.sort_values().iloc[:30].index\n",
    "\n",
    "ax.scatter(list(conditional_ranks.loc[k].values),\n",
    "            list(pearson_ranks.loc[k].values))\n",
    "ax.scatter(list(conditional_ranks.loc[ridx].values),\n",
    "            list(pearson_ranks.loc[ridx].values), label='rhamnolipids', c='r')\n",
    "ax.plot(np.linspace(0,30), np.linspace(0,30), c='r')\n",
    "ax.set_xlabel('Conditional ranks', fontsize=18)\n",
    "ax.set_ylabel('Pearson ranks', fontsize=18)\n",
    "\n",
    "ax.xaxis.set_tick_params(labelsize=16)\n",
    "ax.yaxis.set_tick_params(labelsize=16)\n",
    "ax.xaxis.set_major_locator(MaxNLocator(4))\n",
    "ax.yaxis.set_major_locator(MaxNLocator(4))\n",
    "\n",
    "plt.legend(fontsize=18)\n",
    "plt.tight_layout()\n",
    "plt.savefig('../results/rhamnolipid_ranks.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below are some stats for Pearson correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# annotated compounds known to be produced by Pseudomonas\n",
    "compounds = [\n",
    "    'Pyochelin', 'Phenazine-1-carboxylic Acid', \n",
    "    'HHQ', 'NHQ', 'C9:db-PQS, C9:DB-NQNO', \n",
    "    'PQS/HQNO', 'Pyocyanin'\n",
    "]\n",
    "compound_idx = [ms_md.loc[c, 'shorthand'] in compounds for c in ms_md.index]\n",
    "\n",
    "print('Total detected with Pearson with FDR correction', \n",
    "      multipletests(pear_pvals, method='fdr_bh')[0].sum())\n",
    "\n",
    "print('Annotated compounds detected with Pearson without FDR correction', \n",
    "      (pear_pvals[compound_idx] < 0.05).sum())\n",
    "\n",
    "print('Annotated compounds detected with Pearson with FDR correction', \n",
    "      (multipletests(pear_pvals, method='fdr_bh')[0][compound_idx]).sum())\n",
    "\n",
    "print('Number of Rhamnolipids detected by Pearson without FDR correction', \n",
    "      (pear_pvals[ridx] < 0.05).sum())\n",
    "\n",
    "print('Number of Rhamnolipids detected by Pearson with FDR correction', \n",
    "      pd.Series((multipletests(pear_pvals, method='fdr_bh')[0]), \n",
    "                index=pear_pvals.index).loc[ridx].sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To look in more detail about the specific correlations, we can look at the P. aeruginosa abundance vs Pyocyanin (which was quite different between the two)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "j = 'X211.0918mz161.2083' # pyocyanin\n",
    "ax.scatter(list(microbes[pseudomonas].values),\n",
    "            list(metabolites[j].values))\n",
    "ax.set_xlabel(r'$\\it{P. aeruginosa}$ proportion', fontsize=18)\n",
    "ax.set_ylabel('Pyocyanin proportion', fontsize=18)\n",
    "ax.xaxis.set_tick_params(labelsize=16)\n",
    "ax.yaxis.set_tick_params(labelsize=16)\n",
    "ax.xaxis.set_major_locator(MaxNLocator(4))\n",
    "ax.yaxis.set_major_locator(MaxNLocator(4))\n",
    "plt.tight_layout()\n",
    "plt.savefig('../results/pyocyanin_P.aeruginosa.png')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
